runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java

```diff
@@ -237,13 +245,22 @@ public IndexMapping<K, V> cleanUp(@Nonnull K key, @Nonnull V value) {
             return new IndexMapping<>(updatedMapping, index);
         }
 
-        public IndexMapping<K, V> update(@Nonnull K key, @Nonnull V value) {
-            Map<Object, Map<K, V>> updatedMapping = mapping;
+        public IndexMapping<K, V> update(@Nonnull K key, @Nonnull V value,
+                                         Option<HashArrayMappedTrieModule.LeafSingleton<K, V>> leafNode) {
+            HashArrayMappedTrie<Object, HashArrayMappedTrie<K, V>> updatedMapping = mapping;
 
             Iterable<?> mappedValues = index.getMultiValueIndexFunction().apply(key, value);
             for (Object indexKey: mappedValues) {
-                final Map<K, V> slot = updatedMapping.getOrElse(indexKey, HashMap.empty()).put(key, value);
+                //do mainmap.get(key) to get the leafSingleton class
+                HashArrayMappedTrie<K, V> slot;
+                if (!leafNode.isEmpty()) {
+                    slot = updatedMapping.getOrElse(indexKey, HashArrayMappedTrie.empty()).putNode(leafNode.get());
+                    cacheReuseCounter.ifPresent(Counter::increment);
+                } else {
+                    slot = updatedMapping.getOrElse(indexKey, HashArrayMappedTrie.empty()).put(key, value);
+                }
```

Comment:
Under what conditions can this optional leafNode be empty?

Reply:
A LeafList node returns empty because we don't support getNode from a LeafList node